src/main/java/org/jenkinsci/plugins/KeycloakAccess.java

```diff
@@ -228,18 +229,77 @@ public UserDetails authenticate( String username, String password ) throws Authe
-				Configuration configuration = new Configuration( keycloakDeployment.getAuthServerBaseUrl(), keycloakDeployment.getRealm(), keycloakDeployment.getResourceName(), keycloakDeployment.getResourceCredentials(), keycloakDeployment.getClient() );
-				AuthzClient authzClient = AuthzClient.create( configuration );
-				AccessTokenResponse accessTokenResponse = authzClient.obtainAccessToken();
-				token = accessTokenResponse.getToken();
-				LOGGER.finest( "Token expires in: " + accessTokenResponse.getExpiresIn() );
-				cache.setSystemToken( token, accessTokenResponse.getExpiresIn() );
+		String token = null;
+		SystemTokenInfo systemToken = cache.getSystemAccessToken();
+		if (systemToken != null) {
+			long now = System.currentTimeMillis();
+			if ( now > systemToken.getTokenExpirationTime() ) {
+				//Attempt to refresh
+				if ( now <= systemToken.getRefreshTokenExpirationTime() ) {
+					LOGGER.fine("Refreshing token");
+					AccessTokenResponse newToken = refreshToken( systemToken.getAccessToken().getRefreshToken() );
+					if (newToken != null) {
+						LOGGER.finest( "New Token expires in: " + newToken.getExpiresIn() );
+						LOGGER.finest( "New Refresh Token Expires In: " + newToken.getRefreshExpiresIn() );
+
+						systemToken = new SystemTokenInfo( newToken );
+						cache.storeSystemAccessToken( systemToken );
+						token = systemToken.getAccessToken().getToken();
+					} else {
+						systemToken = null;
+					}
+				} else {
+					//Couldn't refresh, need a new token
+					LOGGER.info("Refresh token is expired, need a new token");
+					systemToken = null;
+				}
+			} else {
+				LOGGER.finer("Returning Cached System Token");
+				return systemToken.getAccessToken().getToken();
+			}
+		}
+
+		if (systemToken == null) {
+			LOGGER.info("Retrieving new System Token");
+			Configuration configuration = new Configuration( keycloakDeployment.getAuthServerBaseUrl(), keycloakDeployment.getRealm(), keycloakDeployment.getResourceName(), keycloakDeployment.getResourceCredentials(), keycloakDeployment.getClient() );
+			AuthzClient authzClient = AuthzClient.create( configuration );
+			AccessTokenResponse accessTokenResponse = authzClient.obtainAccessToken();
+			cache.storeSystemAccessToken(new SystemTokenInfo( accessTokenResponse ));
+			LOGGER.finest( "Token expires in: " + accessTokenResponse.getExpiresIn() );
+			LOGGER.finest( "Refresh Token Expires In: " + accessTokenResponse.getRefreshExpiresIn() );
+			token = accessTokenResponse.getToken();
 		}
 		return token;
 	}
 
+	private AccessTokenResponse refreshToken( String refreshToken ) {
+		String url = keycloakDeployment.getAuthServerBaseUrl() + "/realms/" + keycloakDeployment.getRealm() + "/protocol/openid-connect/token";
+		String clientId = keycloakDeployment.getResourceName();
+		String secret = (String)keycloakDeployment.getResourceCredentials().get("secret");
+		Http http = new Http(getKeycloakConfig(), (params, headers) -> {});
+		return http.<AccessTokenResponse>post(url)

```

Comment:
Does this need to be closed or will it free up the resources automatically? I couldn't tell directly from the source ( https://github.com/keycloak/keycloak/blob/master/authz/client/src/main/java/org/keycloak/authorization/client/util/HttpMethod.java#L102 -> https://github.com/keycloak/keycloak/blob/master/authz/client/src/main/java/org/keycloak/authorization/client/util/HttpMethodResponse.java ) so I am guessing it does it itself?

Reply:
It's definitely how they are using it. But I'm going to dig a little more to make sure: 
https://github.com/keycloak/keycloak/blob/master/authz/client/src/main/java/org/keycloak/authorization/client/AuthzClient.java#L194

Reply:
EntityUtils closes the InputStream which is what is needed for the connection to close. https://github.com/keycloak/keycloak/blob/master/authz/client/src/main/java/org/keycloak/authorization/client/util/HttpMethod.java#L88